File: frontend/lib/api/companies.ts
import { apiFetch } from "@/lib/api/client";
import type {
  Company,
  CompanyFilters,
  CompanyListResponse,
  CompanyMeta,
  Deal,
} from "@/types/api";

export type CompanyInput = {
  company_name: string;
  company_code?: string | null;
  industry?: string | null;
  website?: string | null;
  email?: string | null;
  phone?: string | null;
  address?: string | null;
  city?: string | null;
  state?: string | null;
  country?: string | null;
  postal_code?: string | null;
  annual_revenue?: number | null;
  employee_count?: number | null;
  owner_id?: string | null;
  description?: string | null;
};

export const INDUSTRY_LABELS: Record<string, string> = {
  technology: "Technology",
  finance: "Finance",
  healthcare: "Healthcare",
  manufacturing: "Manufacturing",
  retail: "Retail",
  education: "Education",
  real_estate: "Real Estate",
  consulting: "Consulting",
  media: "Media",
  other: "Other",
};

function buildQuery(filters: CompanyFilters): string {
  const params = new URLSearchParams();
  if (filters.q) params.set("q", filters.q);
  if (filters.industry) params.set("industry", filters.industry);
  if (filters.owner_id) params.set("owner_id", filters.owner_id);
  if (filters.city) params.set("city", filters.city);
  if (filters.country) params.set("country", filters.country);
  if (filters.page) params.set("page", String(filters.page));
  if (filters.page_size) params.set("page_size", String(filters.page_size));
  if (filters.sort_by) params.set("sort_by", filters.sort_by);
  if (filters.sort_order) params.set("sort_order", filters.sort_order);
  const query = params.toString();
  return query ? `?${query}` : "";
}

export async function getCompanyMeta(slug: string): Promise<CompanyMeta> {
  return apiFetch<CompanyMeta>(`/tenants/${slug}/companies/meta`);
}

export async function listCompanies(
  slug: string,
  filters: CompanyFilters = {},
): Promise<CompanyListResponse> {
  return apiFetch<CompanyListResponse>(`/tenants/${slug}/companies${buildQuery(filters)}`);
}

export async function getCompany(slug: string, companyId: string): Promise<Company> {
  return apiFetch<Company>(`/tenants/${slug}/companies/${companyId}`);
}

export async function createCompany(slug: string, data: CompanyInput): Promise<Company> {
  return apiFetch<Company>(`/tenants/${slug}/companies`, {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export async function updateCompany(
  slug: string,
  companyId: string,
  data: Partial<CompanyInput>,
): Promise<Company> {
  return apiFetch<Company>(`/tenants/${slug}/companies/${companyId}`, {
    method: "PATCH",
    body: JSON.stringify(data),
  });
}

export async function deleteCompany(slug: string, companyId: string): Promise<void> {
  await apiFetch<void>(`/tenants/${slug}/companies/${companyId}`, { method: "DELETE" });
}

export async function listCompanyDeals(slug: string, companyId: string): Promise<Deal[]> {
  return apiFetch<Deal[]>(`/tenants/${slug}/companies/${companyId}/deals`);
}

export function formatCompanyLocation(company: Company): string {
  return [company.city, company.state, company.country].filter(Boolean).join(", ") || "—";
}

export function formatCurrency(value: string | null, currency = "USD"): string {
  if (!value) return "—";
  const amount = Number(value);
  if (Number.isNaN(amount)) return "—";
  return new Intl.NumberFormat("en-US", { style: "currency", currency }).format(amount);
}

export function formatDate(value: string): string {
  return new Date(value).toLocaleDateString("en-US", {
    year: "numeric",
    month: "short",
    day: "numeric",
  });
}

File: backend/app/api/v1/companies.py
from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.core.deps import TenantContext, require_permission
from app.db.session import get_db
from app.models.company import COMPANY_INDUSTRIES, COMPANY_SORT_FIELDS
from app.schemas.company import (
    CompanyCreate,
    CompanyListResponse,
    CompanyMetaResponse,
    CompanyResponse,
    CompanyUpdate,
)
from app.schemas.deal import DealResponse
from app.services.company_service import CompanyService, paginate

router = APIRouter(prefix="/tenants/{slug}/companies", tags=["companies"])


def _to_response(company) -> CompanyResponse:
    owner = None
    if company.owner:
        owner = {
            "id": company.owner.id,
            "full_name": company.owner.full_name,
            "email": company.owner.email,
        }
    return CompanyResponse(
        id=company.id,
        tenant_id=company.tenant_id,
        company_name=company.company_name,
        company_code=company.company_code,
        industry=company.industry,
        website=company.website,
        email=company.email,
        phone=company.phone,
        address=company.address,
        city=company.city,
        state=company.state,
        country=company.country,
        postal_code=company.postal_code,
        annual_revenue=company.annual_revenue,
        employee_count=company.employee_count,
        owner_id=company.owner_id,
        owner=owner,
        description=company.description,
        created_by_id=company.created_by_id,
        created_at=company.created_at,
        updated_at=company.updated_at,
    )


@router.get("/meta", response_model=CompanyMetaResponse)
def get_company_meta(
    _: TenantContext = Depends(require_permission("company:read")),
) -> CompanyMetaResponse:
    return CompanyMetaResponse(
        industries=list(COMPANY_INDUSTRIES),
        sort_fields=list(COMPANY_SORT_FIELDS),
    )


@router.get("", response_model=CompanyListResponse)
def list_companies(
    q: str | None = Query(default=None, max_length=200),
    industry: str | None = Query(default=None, max_length=100),
    owner_id: UUID | None = Query(default=None),
    city: str | None = Query(default=None, max_length=100),
    country: str | None = Query(default=None, max_length=100),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    sort_by: str = Query(default="created_at"),
    sort_order: str = Query(default="desc", pattern="^(asc|desc)$"),
    ctx: TenantContext = Depends(require_permission("company:read")),
    db: Session = Depends(get_db),
) -> CompanyListResponse:
    service = CompanyService(db)
    companies, total = service.list_companies(
        ctx.tenant.id,
        q=q,
        industry=industry,
        owner_id=owner_id,
        city=city,
        country=country,
        page=page,
        page_size=page_size,
        sort_by=sort_by,
        sort_order=sort_order,
    )
    meta = paginate(total, page, page_size)
    return CompanyListResponse(
        items=[_to_response(company) for company in companies],
        **meta,
    )


@router.post("", response_model=CompanyResponse, status_code=201)
def create_company(
    payload: CompanyCreate,
    ctx: TenantContext = Depends(require_permission("company:write")),
    db: Session = Depends(get_db),
) -> CompanyResponse:
    company = CompanyService(db).create_company(
        ctx.tenant.id, payload, ctx.membership.user_id
    )
    return _to_response(company)


@router.get("/{company_id}", response_model=CompanyResponse)
def get_company(
    company_id: UUID,
    ctx: TenantContext = Depends(require_permission("company:read")),
    db: Session = Depends(get_db),
) -> CompanyResponse:
    company = CompanyService(db).get_company(ctx.tenant.id, company_id)
    return _to_response(company)


@router.get("/{company_id}/deals", response_model=list[DealResponse])
def list_company_deals(
    company_id: UUID,
    ctx: TenantContext = Depends(require_permission("company:read")),
    db: Session = Depends(get_db),
) -> list[DealResponse]:
    from app.api.v1.deals import _to_response as deal_to_response

    deals = CompanyService(db).list_company_deals(ctx.tenant.id, company_id)
    return [deal_to_response(deal) for deal in deals]


@router.patch("/{company_id}", response_model=CompanyResponse)
def update_company(
    company_id: UUID,
    payload: CompanyUpdate,
    ctx: TenantContext = Depends(require_permission("company:write")),
    db: Session = Depends(get_db),
) -> CompanyResponse:
    company = CompanyService(db).update_company(ctx.tenant.id, company_id, payload, ctx.membership.user_id)
    return _to_response(company)


@router.delete("/{company_id}", status_code=204)
def delete_company(
    company_id: UUID,
    ctx: TenantContext = Depends(require_permission("company:delete")),
    db: Session = Depends(get_db),
) -> None:
    CompanyService(db).delete_company(ctx.tenant.id, company_id, ctx.membership.user_id)
